src/kvstore/collectives/src/collectives.cc

```diff
@@ -0,0 +1,855 @@
+    record.root_rank = root_rank;
+    record.val_in = v_invals[i];
+    MXCOLL_DEBUG(coll_global.rank, "MXBroadCastImpl insert one record key [%s]!\n",
+                record.key.c_str());
+
+    auto broadcast_async_fn = [record]
+    (mxnet::RunContext rctx, mxnet::Engine::CallbackOnComplete cb) {
+      EnqueueCollective(record, MPIRequest::BROADCAST, cb);
+    };
+    CHECK_NOTNULL(mxnet::Engine::Get())->PushAsync(
+      broadcast_async_fn,
+      coll_global.pinned_ctx,
+      {},
+      {record.val_in->var()},
+      mxnet::FnProperty::kNormal,
+      priority, "KVSTORE BROADCAST");
+  }
+  return 0;
+}
+
+int MXBroadcast(const std::vector<int> &keys,
+                const std::vector<mxnet::NDArray*> &values,
+                int root_rank,
+                int priority) {
+  std::vector<std::string> v_keys;
+  std::string key_prefix  = INT_PREFIX;
+  std::string idx_prefix  = IDX_PREFIX;
+  std::string delimiter   = DELIMITER;
+  std::string ops_prefix  = OPS_PREFIX;
+  std::string ops_broadcast   = OPS_BROADCAST;
+  std::string new_key;
+  size_t idx = 0;
+  for (auto& key : keys) {
+    // To simplify original logic for group key value, we rename the original
+    // duplicated key and make every key unique now.
+    size_t index = countIDX(keys, key, idx);
+    new_key = ops_prefix + delimiter + ops_broadcast + delimiter +
+              key_prefix + delimiter + std::to_string(key) + delimiter +
+              idx_prefix + delimiter + std::to_string(index);
+    v_keys.push_back(new_key);
+    idx++;
+  }
+  return MXBroadcastImpl(v_keys, values, root_rank, priority);
+}
+
+int MXBroadcastEx(const std::vector<std::string> &keys,
+                  const std::vector<mxnet::NDArray*> &values,
+                  int root_rank,
+                  int priority) {
+  std::vector<std::string> v_keys;
+  std::string key_prefix  = STR_PREFIX;
+  std::string idx_prefix  = IDX_PREFIX;
+  std::string delimiter   = DELIMITER;
+  std::string ops_prefix  = OPS_PREFIX;
+  std::string ops_broadcast   = OPS_BROADCAST;
+  std::string new_key;
+  size_t idx = 0;
+  for (auto& key : keys) {
+    // To simplify original logic for group key value pairs, we rename the original
+    // duplicated key and make every key unique now.
+    size_t index = countIDX(keys, key, idx);
+    new_key = ops_prefix + delimiter + ops_broadcast + delimiter +
+              key_prefix + delimiter + key + delimiter +
+              idx_prefix + delimiter + std::to_string(index);
+    v_keys.push_back(new_key);
+    idx++;
+  }
+  return MXBroadcastImpl(v_keys, values, root_rank, priority);
+}
+
+int MXAllGather(const std::vector<int> &keys,
+                const std::vector<mxnet::NDArray*> &values,
+                int priority) {
+  // place holder
+  return 0;
+}
+
+int MXAllGatherEx(const std::vector<std::string> &keys,
+                  const std::vector<mxnet::NDArray*> &values,
+                  int priority) {
+  // place holder
```

Comment:
If it's not implemented, could you raise an error instead of returning 0 silently?

Reply:
OK.